File: app/src/main/java/com/vwoom/timelapsegallery/utils/FileUtils.java
package com.vwoom.timelapsegallery.utils;

import android.content.Context;
import android.os.Environment;
import android.util.Log;

import com.vwoom.timelapsegallery.database.entry.PhotoEntry;
import com.vwoom.timelapsegallery.database.entry.ProjectEntry;
import com.vwoom.timelapsegallery.database.view.Project;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

public final class FileUtils {

    public static final String ReservedChars = "|\\?*<\":>+[]/'";
    public static final String TEMP_FILE_SUBDIRECTORY = "temporary_images";
    private static final String TAG = FileUtils.class.getSimpleName();

    /* Used to create a photo file in its final location */
    private static File createImageFileForProject(File storageDirectory, ProjectEntry currentProject, long timestamp) {
        // Create an image file name from the current timestamp
        String imageFileName = timestamp + ".jpg";
        File projectDir = getProjectFolder(storageDirectory, currentProject);

        if (!projectDir.exists()) projectDir.mkdirs();

        return new File(projectDir, imageFileName);
    }

    private static File getProjectFolder(File externalFilesDir, ProjectEntry project){
        //File storageDir = context.getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        String projectPath = getProjectDirectoryPath(project);
        return new File(externalFilesDir, projectPath);
    }

    private static File getProjectFolder(File externalFilesDir, Project project){
        //File storageDir = context.getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        String projectPath = getProjectDirectoryPath(project);
        return new File(externalFilesDir, projectPath);
    }

    public static List<PhotoEntry> getPhotosInDirectory(File externalFilesDir, ProjectEntry project){
        List<PhotoEntry> photos = new ArrayList<>();
        File projectFolder = getProjectFolder(externalFilesDir, project);
        File[] files = projectFolder.listFiles();

        long projectId = project.getId();
        if (files != null) {
            for (File child : files) {
                String url = child.getAbsolutePath();
                String filename = url.substring(url.lastIndexOf("/")+1);
                long timestamp = Long.valueOf(filename.replaceFirst("[.][^.]+$",""));
                PhotoEntry photoEntry = new PhotoEntry(projectId, timestamp);
                photos.add(photoEntry);
            }
        } else return null;

        return photos;
    }

    /* Creates a file in a temporary location */
    public static File createTemporaryImageFile(File externalFilesDir) throws IOException {
        // Create an image file name
        String imageFileName = "TEMP_";
        //File storageDir = context.getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        File tempFolder = new File(externalFilesDir, TEMP_FILE_SUBDIRECTORY);
        if (!tempFolder.exists()) tempFolder.mkdirs();

        return File.createTempFile(
                imageFileName,  /* prefix */
                ".jpg",         /* suffix */
                tempFolder      /* directory */
        );
    }

    /* Used to create a photo file from its temporary location */
    public static File createFinalFileFromTemp(
            File externalFilesDir,
            String tempPath,
            ProjectEntry currentProject,
            long timestamp) throws IOException {
        // Create the permanent file for the photo
        File finalFile = createImageFileForProject(externalFilesDir, currentProject, timestamp);
        // Create tempfile from previous path
        File tempFile = new File(tempPath);
        // Copy file to new destination
        copy(tempFile, finalFile);
        // Remove temporary file
        tempFile.delete();

        return finalFile;
    }

    /* Used to copy from one file to another */
    private static void copy(File src, File dst) throws IOException {
        InputStream in = new FileInputStream(src);
        try {
            OutputStream out = new FileOutputStream(dst);
            try{
                byte[] buf = new byte[1024];
                int len;
                while ((len=in.read(buf)) > 0) {
                    out.write(buf, 0, len);
                }
            } finally {
                out.close();
            }
        } finally {
            in.close();
        }
    }

    /* Copies a Project from one folder to another */
    public static boolean renameProject(Context context, ProjectEntry source, ProjectEntry destination){
        // Create a file for the source project
        File storageDir = context.getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        String sourceProjectPath = getProjectDirectoryPath(source);
        File sourceProject = new File(storageDir, sourceProjectPath);

        // Create a file for the destination project
        String destinationProjectPath = getProjectDirectoryPath(destination);
        File destinationProject = new File(storageDir, destinationProjectPath);

        // Rename the folder
        boolean success = sourceProject.renameTo(destinationProject);

        // Update the photo references
        // Return true if rename is successful
        if (success) return true;
        // Return false if rename is not successful
        return false;
    }

    /* Delete temporary directory and files within temporary directory */
    public static void deleteTempFiles(Context context){
        File storageDir = context.getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        File tempDir = new File(storageDir, TEMP_FILE_SUBDIRECTORY);
        deleteRecursive(tempDir);
        Log.d("deletion check", "delete temp files firing");
    }

    /* Recursive delete used by delete project, delete temp, or delete photo */
    private static void deleteRecursive(File fileOrFileDirectory){
        if (fileOrFileDirectory.isDirectory()){
            for (File child : fileOrFileDirectory.listFiles()){
                deleteRecursive(child);
            }
        }
        Log.d("deletion check", "deleting temporary file named: " + fileOrFileDirectory.getName());
        fileOrFileDirectory.delete();
    }

    /* Delete project directory and files within project directory */
    public static void deleteProject(Context context, ProjectEntry projectEntry){
        String projectDirectoryPath = getProjectDirectoryPath(projectEntry);
        File storageDir = context.getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        File projectDirectory = new File(storageDir, projectDirectoryPath);
        deleteRecursive(projectDirectory);
    }

    /* Deletes file referred to in photo entry */
    public static void deletePhoto(File externalFilesDir, ProjectEntry projectEntry, PhotoEntry photoEntry){
        File photoFile = new File(getPhotoUrl(externalFilesDir, projectEntry, photoEntry));
        deleteRecursive(photoFile);
    }
    public static void deletePhoto(File externalFilesDir, Project project, PhotoEntry photoEntry){
        File photoFile = new File(getPhotoUrl(externalFilesDir, project, photoEntry));
        deleteRecursive(photoFile);
    }

    /* Returns true if a path contains reserved characters */
    public static boolean pathContainsReservedCharacter(String path){
        for (int i = 0; i < ReservedChars.length(); i++){
            char current = ReservedChars.charAt(i);
            if (path.indexOf(current) >= 0) return true;
        }
        return false;
    }

    /* Returns the pattern for a projects path : project path = {project_id}_{project_name}
    * example: 1_My Project
    * */
    private static String getProjectDirectoryPath(ProjectEntry projectEntry){
        String name = projectEntry.getProject_name();
        if (name == null) return String.valueOf(projectEntry.getId());
        else return projectEntry.getId() + "_" + projectEntry.getProject_name();
    }

    private static String getProjectDirectoryPath(Project project){
        if (project.getProject_name() == null) return String.valueOf(project.getProject_id());
        else return project.getProject_id() + "_" + project.getProject_name();
    }

    public static String getPhotoUrl(File externalFilesDir, ProjectEntry projectEntry, PhotoEntry photoEntry){
        String imageFileName = getPhotoFileName(photoEntry);
        File projectDir = getProjectFolder(externalFilesDir, projectEntry);
        File photoFile = new File(projectDir, imageFileName);
        return photoFile.getAbsolutePath();
    }

    public static String getPhotoUrl(File externalFilesDir, Project project, PhotoEntry photoEntry){
        String imageFileName = getPhotoFileName(photoEntry);
        File projectDir = getProjectFolder(externalFilesDir, project);
        File photoFile = new File(projectDir, imageFileName);
        return photoFile.getAbsolutePath();
    }

    public static String getPhotoUrl(File externalFilesDir, Project project){
        String imageFileName = getPhotoFileName(project.getCover_photo_timestamp());
        File projectDir = getProjectFolder(externalFilesDir, project);
        File photoFile = new File(projectDir, imageFileName);
        return photoFile.getAbsolutePath();
    }

    public static String getPhotoFileName(PhotoEntry entry){
        return entry.getTimestamp() + ".jpg";
    }

    public static String getPhotoFileName(long timestamp){
        return timestamp + ".jpg";
    }
}

File: app/src/main/java/com/vwoom/timelapsegallery/utils/FileUtils.kt
package com.vwoom.timelapsegallery.utils

import com.vwoom.timelapsegallery.data.entry.PhotoEntry
import com.vwoom.timelapsegallery.data.entry.ProjectEntry
import com.vwoom.timelapsegallery.data.entry.ProjectScheduleEntry
import com.vwoom.timelapsegallery.data.entry.TagEntry
import com.vwoom.timelapsegallery.utils.ProjectUtils.getMetaDirectoryForProject
import com.vwoom.timelapsegallery.utils.ProjectUtils.getProjectFolder
import timber.log.Timber
import java.io.*

object FileUtils {

    private const val RESERVED_CHARACTERS = "|\\?*<\":>+[]/'"

    // Directory definitions
    private const val TEMP_FILE_SUBDIRECTORY = "temporary_images"
    private const val PROJECTS_FILE_SUBDIRECTORY = "Projects"
    private const val META_FILE_SUBDIRECTORY = "Meta"
    private const val GIF_FILE_SUBDIRECTORY = "Gif"

    // Text files for metadata
    private const val SCHEDULE_DEFINITION_TEXT_FILE = "schedule.txt"
    private const val TAGS_DEFINITION_TEXT_FILE = "tags.txt"
    private const val SENSOR_DEFINITION_TEXT_FILE = "sensorData.txt"
    private const val LIST_PHOTOS_TEXT_FILE = "photosList.txt"

    fun getReservedCharacters() = RESERVED_CHARACTERS

    fun getProjectsSubdirectory(externalFilesDir: File): File {
        return File(externalFilesDir, PROJECTS_FILE_SUBDIRECTORY)
    }

    fun getMetaSubdirectory(externalFilesDir: File): File {
        return File(externalFilesDir, META_FILE_SUBDIRECTORY)
    }

    fun getGifSubdirectory(externalFilesDir: File): File {
        return File(externalFilesDir, GIF_FILE_SUBDIRECTORY)
    }

    fun getTempFolder(externalFilesTestDir: File): File {
        return File(externalFilesTestDir, TEMP_FILE_SUBDIRECTORY)
    }


    fun getTagsFile(projectMetaDir: File): File {
        return File(projectMetaDir, TAGS_DEFINITION_TEXT_FILE)
    }

    fun getScheduleFile(projectMetaDir: File): File{
        return File(projectMetaDir, SCHEDULE_DEFINITION_TEXT_FILE)
    }

    fun getSensorFile(projectMetaDir: File): File {
        return File(projectMetaDir, SENSOR_DEFINITION_TEXT_FILE)
    }

    // Creates an empty image file for a project in the projects folder to be copied to
    private fun createImageFileForProject(externalFilesDir: File, projectEntry: ProjectEntry, timestamp: Long): File {
        val imageFileName = "$timestamp.jpg"
        val projectDir = getProjectFolder(externalFilesDir, projectEntry)
        if (!projectDir.exists()) projectDir.mkdirs()
        return File(projectDir, imageFileName)
    }

    // Creates an empty image file in the temporary directory to be written to by the camera
    @JvmStatic
    @Throws(IOException::class)
    fun createImageFileInTemporaryFolder(externalFilesDir: File?): File {
        val imageFilePrefix = "TEMP_"
        val suffixExtension = ".jpg"
        val temporaryFolder = File(externalFilesDir, TEMP_FILE_SUBDIRECTORY)
        if (!temporaryFolder.exists()) temporaryFolder.mkdirs()
        return File.createTempFile(
                imageFilePrefix,  // prefix
                suffixExtension,  // suffix
                temporaryFolder // directory
        )
    }

    // Copies from temporary photo file to the final photo for a project
    @Throws(IOException::class)
    fun createProjectPhotoFileFromTemporaryPhoto(
            externalFilesDir: File,
            tempPath: String,
            projectEntry: ProjectEntry,
            timestamp: Long): File {
        // 1. First create the permanent file for the photo
        val finalFile = createImageFileForProject(externalFilesDir, projectEntry, timestamp)
        // 2. Get reference to the temp file
        val tempFile = File(tempPath)
        // 3. Copy from temp to the final destination
        copyFile(tempFile, finalFile)
        // 4. Clean up / remove the temporary file
        tempFile.delete()
        return finalFile
    }

    // Used internal to class to copy from one file to another (should be from the temp photo to final photo)
    @Throws(IOException::class)
    private fun copyFile(src: File, destination: File) {
        val input: InputStream = FileInputStream(src)
        input.use {
            val out: OutputStream = FileOutputStream(destination)
            out.use {
                val buf = ByteArray(1024)
                var len: Int
                while (input.read(buf).also { len = it } > 0) {
                    out.write(buf, 0, len)
                }
            }
        }
    }

    // Deletes the temporary directory and its contents
    fun deleteTempFiles(externalFilesDir: File?) {
        if (externalFilesDir == null) return
        val tempDir = File(externalFilesDir, TEMP_FILE_SUBDIRECTORY)
        deleteRecursive(tempDir)
    }

    // Recursive deletion of a file or directory
    fun deleteRecursive(fileOrFileDirectory: File) {
        if (fileOrFileDirectory.isDirectory) {
            val files = fileOrFileDirectory.listFiles()
            if (files != null)
                for (child in files) {
                    deleteRecursive(child)
                }
        }
        fileOrFileDirectory.delete()
    }

    // Returns true if a path contains a reserved character
    fun pathContainsReservedCharacter(path: String): Boolean {
        for (character in RESERVED_CHARACTERS) {
            if (path.contains(character)) return true
        }
        return false
    }

    // Supports .jpg, .png, .jpeg creation from timestamp
    fun getPhotoFileExtensions(timestamp: Long): Array<String> {
        return arrayOf("$timestamp.jpg", "$timestamp.png", "$timestamp.jpeg")
    }

    // This function Writes the tags for a project in a text file in the meta directory for the project
    // Ex. a project of ID 1 will have a path of ...meta/1/tags.txt
    // In general this is used to decouple tag data from the room data (for use in loading data on and off devices)
    fun writeProjectTagsFile(
            externalFilesDir: File,
            projectId: Long,
            // TODO (1.4) : look into preserving tag order so that lineage can be preserved for plants (mother x father designation). Currently tags are sorted alphabetically somewhere
            tags: List<TagEntry>) {
        val metaDir = getMetaDirectoryForProject(externalFilesDir, projectId)
        val tagsFile = File(metaDir, TAGS_DEFINITION_TEXT_FILE)

        // Write the tags to a text file
        try {
            val output = FileOutputStream(tagsFile)
            val outputStreamWriter = OutputStreamWriter(output)
            for (tag in tags) {
                outputStreamWriter.write(tag.text + "\n")
            }
            outputStreamWriter.flush()
            output.fd.sync()
            outputStreamWriter.close()
        } catch (exception: IOException) {
            Timber.e("error writing tag to text file: ${exception.message}")
        }
    }

    // Writes the schedule for a project in the projects meta directory similarly to the tag writing function
    // This text file should contain a simple integer indicating the interval between scheduled photos in days
    // (For use in loading data on and off devices)
    fun writeProjectScheduleFile(
            externalFilesDir: File,
            projectId: Long,
            projectScheduleEntry: ProjectScheduleEntry) {
        val metaDir = getMetaDirectoryForProject(externalFilesDir, projectId)
        val scheduleFile = File(metaDir, SCHEDULE_DEFINITION_TEXT_FILE)

        try {
            val output = FileOutputStream(scheduleFile)
            val outputStreamWriter = OutputStreamWriter(output)
            outputStreamWriter.write(projectScheduleEntry.interval_days.toString())
            outputStreamWriter.flush()
            output.fd.sync()
            outputStreamWriter.close()
        } catch (exception: IOException) {
            Timber.e("error writing schedule to text file: ${exception.message}")
        }
    }

    // Creates a temporary text file which contains a list of the photo urls for a project
    // This text file is used to control ffmpeg in order to create a GIF from the set of photos
    // The format required by ffmpeg is as follows:
    // file '/path/to/file1'
    // file '/path/to/file2'
    // etc.
    fun createTempListPhotoFiles(
            externalFilesDir: File,
            projectEntry: ProjectEntry): File? {
        // 1. First clean the temp files
        deleteTempFiles(externalFilesDir)

        // 2. Get the list of photos to convert
        val listOfPhotos = ProjectUtils.getPhotoEntriesInProjectDirectory(externalFilesDir, projectEntry)

        // 3. Create the temporary folder and define the text file
        val tempFolder = File(externalFilesDir, TEMP_FILE_SUBDIRECTORY)
        tempFolder.mkdir()
        val listFiles = File(tempFolder, LIST_PHOTOS_TEXT_FILE)

        // Write the file paths to the text file
        try {
            val output = FileOutputStream(listFiles)
            val outputStreamWriter = OutputStreamWriter(output)
            // For each photo in the list write its path
            for (photo in listOfPhotos) {
                val photoUrlString = ProjectUtils.getProjectPhotoUrl(externalFilesDir, projectEntry, photo.timestamp)
                val fileDefString = "file '$photoUrlString'\n" // file '/path/to/file1'
                outputStreamWriter.write(fileDefString)
            }
            outputStreamWriter.flush()
            output.fd.sync()
            outputStreamWriter.close()
        }
        catch (exception: IOException) {
            return null
        }
        return listFiles
    }

    // Writes sensor data (if available) for each photo to a text file
    // Should be a list of strings in the following format:
    // timestamp(long) light(lx) temperature(celsius) pressure(mbar) humidity(%)
    fun writeSensorData(externalFilesDir: File, photoEntry: PhotoEntry, projectId: Long) {
        val metaDir = getMetaDirectoryForProject(externalFilesDir, projectId)
        val sensorDataFile = File(metaDir, SENSOR_DEFINITION_TEXT_FILE)

        val entryText = "${photoEntry.timestamp} ${photoEntry.light} ${photoEntry.temp} ${photoEntry.pressure} ${photoEntry.humidity}\n"

        try {
            sensorDataFile.appendText(entryText)
        } catch (exception: IOException){
            Timber.e("error writing sensor data to text file: ${exception.message}")
        }
    }

}